[4.0] Remove Meta Keywords (part 1)

brianteeman opened the pull request:
Meta Keywords as a meta tag has pretty much been irrelevant for **10 years** (see https://yoast.com/meta-keywords/  for example)

This PR does the following

1. Removes Site Meta Keywords from global config
2. Removes Menu item Meta Keywords
3. Removes per language Meta Keywords
4. Renames the item field Meta Keywords to Keywords
 - We use this as a field with mod_related_items and com_banners

Part 2 of removing meta keywords will
1. Move the Keyword field to a more appropriate place
2. Check that all the keyword fields are used with mod_related_items and com_banners

### Testing
Check that Meta Keyword field is removed from global config and menu items
Check that renamed Keyword field does not produce a meta key in the frontend

Quy commented on libraries/src/Table/CoreContent.php:
```diff
@@ -145,36 +145,6 @@ public function check()
-			$bad_characters = array("\n", "\r", "\"", '<', '>');
-
-			// Remove bad characters
-			$after_clean = StringHelper::str_ireplace($bad_characters, '', $this->core_metakey);
```
Remove `use Joomla\String\StringHelper;`

Quy commented on libraries/src/Table/Content.php:
```diff
@@ -272,36 +272,6 @@ public function check()
-			$bad_characters = array("\n", "\r", "\"", '<', '>');
-
-			// Remove bad characters
-			$after_clean = StringHelper::str_ireplace($bad_characters, '', $this->metakey);
```
Remove `use Joomla\String\StringHelper;`

HLeithner commented:
Why do you think we should remove it? Who says it's irrelevant in all cases? Is this true for all search engines or only for google?

Our own search engine (smart search) uses the metakey for the ranking...

brianteeman commented:
@HLeithner 
>Why do you think we should remove it? Who says it's irrelevant in all cases? Is this true for all search engines or only for google?

https://yoast.com/meta-keywords/ neither yahoo, google or bing use it

> Our own search engine (smart search) uses the metakey for the ranking...

As stated in the PR
 > Renames the item field Meta Keywords to Keywords

brianteeman commented:
@HLeithner This was also part of the specification for the GSOC 19 SEO project https://docs.google.com/document/d/1MaOtEpp0Ma-e_d7xuXVAnGXP72ec9Ar1XrKtPQvnw4I/edit?usp=sharing

HLeithner commented:
Adding the rest would be much nicer then removing something thats maybe not used ;-) but ok if george or the other maintainers are happy with it i'm too.

brianteeman commented:
> Adding the rest would be much nicer


![yourfreedomcoach-Mervyn-Barrett-constant-improvement](https://user-images.githubusercontent.com/1296369/59695007-6b935e80-91e1-11e9-8def-03bfd1e24d5e.png)

simbus82 commented:
Yoast is the worst source for "SEO best practice". Rubbish for wordpress newbies.

https://yandex.com/support/webmaster/controlling-robot/html.html
https://www.searchenginejournal.com/yandex-seo-guide/252885/

Meta Tag Keywords are important for SEO in Russia.
![image](https://user-images.githubusercontent.com/1375730/59710002-5aaa1380-9208-11e9-9d94-9ef26397c571.png)

It is not a problem to use some plugin or some custom fields to obtain this for russian websites.
And I am the first to tell my clients not to waste time filling in that data. But only to Italian, English, French, etc. customers ... Not to Russian customers!
But Joomla needs a SEO team leader who should have a little more experience in international projects. By the way ... is there a SEO team leader in the organization?

alikon commented:
> But Joomla needs a SEO team leader who should have a little more experience in international projects. 

@simbus82  i understand this like a candidature to do this work, 
and imho we really need people that want to do work .....   

so please submit a PR or a RFC ....whatever

p.s.
italians do it better

brianteeman commented:
The reason for posting the toast link was to save time posting one link and not three. I am just submitting a pr to match the reques I have no clue on SEO.

Quy commented on administrator/language/en-GB/en-GB.com_banners.ini:
```diff
@@ -74,7 +74,7 @@ COM_BANNERS_FIELD_BASENAME_LABEL="File Name"
 COM_BANNERS_FIELD_CLIENT_LABEL="Client"
-COM_BANNERS_FIELD_CLIENT_METAKEYWORDPREFIX_DESC="When matching Meta keywords, only search for Meta keywords with this prefix (improves performance)."
+COM_BANNERS_FIELD_CLIENT_METAKEYWORDPREFIX_DESC="When matching Keywords, only search for Keywords with this prefix (improves performance)."
 COM_BANNERS_FIELD_CLIENT_METAKEYWORDPREFIX_LABEL="Meta Keyword Prefix"
```
```suggestion
COM_BANNERS_FIELD_CLIENT_METAKEYWORDPREFIX_LABEL="Keyword Prefix"
```

Quy commented on administrator/language/en-GB/en-GB.com_banners.ini:
```diff
@@ -89,7 +89,7 @@ COM_BANNERS_FIELD_HEIGHT_LABEL="Height"
 COM_BANNERS_FIELD_IMPTOTAL_LABEL="Max. Impressions"
-COM_BANNERS_FIELD_METAKEYWORDPREFIX_DESC="When matching Meta keywords, only search for Meta keywords with this prefix (improves performance)."
+COM_BANNERS_FIELD_METAKEYWORDPREFIX_DESC="When matching Keywords, only search for Keywords with this prefix (improves performance)."
 COM_BANNERS_FIELD_METAKEYWORDPREFIX_LABEL="Meta Keyword Prefix"
```
```suggestion
COM_BANNERS_FIELD_METAKEYWORDPREFIX_LABEL="Keyword Prefix"
```

Quy commented on language/en-GB/en-GB.mod_banners.ini:
```diff
@@ -13,7 +13,7 @@ MOD_BANNERS_FIELD_COUNT_LABEL="Count"
 MOD_BANNERS_FIELD_RANDOMISE_LABEL="Randomise"
-MOD_BANNERS_FIELD_TAG_DESC="Banner is selected by matching the banner tags to the current document meta keywords."
+MOD_BANNERS_FIELD_TAG_DESC="Banner is selected by matching the banner tags to the current document Keywords."
 MOD_BANNERS_FIELD_TAG_LABEL="Select by Meta Keyword"
```
```suggestion
MOD_BANNERS_FIELD_TAG_LABEL="Select by Keyword"
```

wilsonge commented on libraries/src/Table/CoreContent.php:
```diff
@@ -145,36 +144,6 @@ public function check()
 			$this->core_publish_down = $temp;
 		}
 
-		// Clean up keywords -- eliminate extra spaces between phrases
```
This one we probably still need - anything 3rd party using UCM might be using this (at least JED) and whilst we can make the b/c break I'm not sure it's *that* important

brianteeman replied:
done

wilsonge commented on libraries/src/Table/Content.php:
```diff
@@ -272,36 +271,6 @@ public function check()
 			$this->modified = $this->_db->getNullDate();
 		}
 
-		// Clean up keywords -- eliminate extra spaces between phrases
```
This is the Table instance for com_content - we're keeping the key's here for the keywords field so this is required for sure

brianteeman replied:
Do we really need the cleanup? Happy to put it back if we do

wilsonge replied:
I think it's safer. Because otherwise we're going to end up having issues with search which expects the data in a certain format

brianteeman replied:
done

wilsonge commented:
I've edited the yoast url into the PR description so anyone coming to this PR sees it first :)

wilsonge commented:
@simbus82 you don't need meta keywords for yandex if you're using json ld which we will be once I finish up https://github.com/joomla/joomla-cms/pull/25117 (yandex has been supporting that program since 2011)

wilsonge commented:
We're going to need to document a b/c break to banners if/when we go with this. It seems to filter by the documents keywords. So things like the menu and global config metadata as fallbacks will be relevant there. https://github.com/joomla/joomla-cms/blob/staging/modules/mod_banners/helper.php#L35

In the mean time I'm going to try and ping some people in the Russian community to gain some extra clarification over Yandex

brianteeman commented:
Re Yandex it does use keywords if this is up to date https://yandex.com/support/webmaster/controlling-robot/html.html

atpatil11 commented:
Getting the following error while applying patch.

The file marked for modification does not exist: components/com_search/View/Search/HtmlView.php![screen shot 2019-11-02 at 06 22 09](https://issues.joomla.org/uploads/1/895a7e2a30c14022760d52b9f3075b8e.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

Quiviro commented:
I have tested this item :red_circle: unsuccessfully on a6409db7d40cef895d8b5e1013c5036897ee0257<br /><br />Same error message than [alpatil11](https://issues.joomla.org/tracker/joomla-cms/25258#event-485633 "").
Meta Keywords field still in global config and menu item<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

b2z commented:
Correct me if I am wrong. Updating from J3 to J4 will affect only this - it won't produce a meta key in the frontend?

A little poll by me in Russian Telegram chat about keywords states that keywords are 99.9% useless for Yandex.

wilsonge commented:
Correct

brianteeman commented:
So do I _spend my time_ updating this pr to resolve the conflicts or not?

b2z commented:
Here is one experience with keywords and Yandex. Quoted and translated from Russian:

For a long time I was convinced that keyword is a rudiment, about a year ago I read an article about an experiment where a person added phrases to keywords that were not on the page, and after that he received visibility through them after the  time. Believe it or not - this is, as usual in our topic, everyone’s business, but I personally leave it to myself.

Quy commented:
Unable to install your branch.

> Error
Column count doesn't match value count at row 1

brianteeman commented:
I think I messed up fixing the conflicts

b2z commented:
Just leaving a note - people keep saying that keywords for Yandex is almost useless and for those who want to keep them, can create a plugin :)

sanek4life commented:
@b2z Maybe it will be a **official plugin**, from the main Joomla directory, as it was done for the weblink component?

@brianteeman I mean, it can be moved as a separate plugin here: [https://extensions.joomla.org/category/official-extensions/](https://extensions.joomla.org/category/official-extensions/)

b2z commented:
Well, IMO keywords does not deserve such attention in 2020.

brianteeman commented:
@sanek4life no - either keep it or kill it

wilsonge commented on installation/sql/postgresql/joomla.sql:
```diff
@@ -1368,7 +1368,7 @@ SELECT 19, 'main', 'mod_menu_fields', 'Contact Custom Fields', '', 'contact/Cust
 SELECT 20, 'main', 'mod_menu_fields_group', 'Contact Custom Fields Group', '', 'contact/Custom Fields Group', 'index.php?option=com_fields&view=groups&context=com_contact.contact', 'component', 1, 7, 2, "extension_id", 0, NULL, 0, 0, 'class:messages-add', 0, '', 18, 19, 0, '*', 1, NULL, NULL FROM "#__extensions" WHERE "name" = 'com_fields';
 INSERT INTO "#__menu" ("id", "menutype", "title", "alias", "note", "path", "link", "type", "published", "parent_id", "level", "component_id", "checked_out", "checked_out_time", "browserNav", "access", "img", "template_style_id", "params", "lft", "rgt", "home", "language", "client_id", "publish_up", "publish_down")
-SELECT 101, 'mainmenu', 'Home', 'home', '', 'home', 'index.php?option=com_content&view=featured', 'component', 1, 1, 1, "extension_id", 0, NULL, 0, 1, '', 0, '{"featured_categories":[""],"layout_type":"blog","num_leading_articles":"1","num_intro_articles":"3","num_columns":"3","num_links":"0","multi_column_order":"1","orderby_pri":"","orderby_sec":"front","order_date":"","show_pagination":"2","sh ...
+
```
assuming this wasn't supposed to be fully removed?

wilsonge commented on administrator/components/com_languages/forms/language.xml:
```diff
@@ -1,8 +1,8 @@
 	<fieldset>
-		<field
-			name="lang_id"
+		<field 
```
Revert this change please

wilsonge commented on libraries/src/Table/Content.php:
```diff
@@ -274,6 +274,12 @@ public function check()
 			$this->publish_down = $temp;
 		}
 
+		// Set modified to null date if not set
```
Not quite sure what went wrong here. But assume this isn't intentional?

brianteeman committed:
[4.0] Remove Meta Keywords (part 1)

Meta Keywords as a meta tag has pretty much been irrelevant for **10 years**

This PR does the following

1. Removes Site Meta Keywords from global config
2. Removes Menu item Meta Keywords
3. Removes per language Meta Keywords
4. Renames the item field Meta Keywords to Keywords
 - We use this as a field with mod_related_items and com_banners

Part 2 of removing meta keywords will
1. Move the Keyword field to a more appropriate place
2. Check that all the keyword fields are used with mod_related_items and com_banners

### Testing
Check that Meta Keyword field is removed from global config and menu items
Check that renamed Keyword field does not produce a meta key in the frontend

brianteeman committed:
.

brianteeman commented:
conflicts all resolved - now I understand cherry pick and force push ;)

Quy commented on libraries/src/Table/Content.php:
```diff
@@ -303,15 +309,6 @@ public function check()
-			$this->metakey = '';
-		}
-
-		if ($this->metadesc === null)
```
Revert?

Quy commented on language/en-GB/mod_related_items.sys.ini:
```diff
@@ -5,4 +5,5 @@
 MOD_RELATED_ITEMS_LAYOUT_DEFAULT="Default"
-MOD_RELATED_XML_DESCRIPTION="This module displays other Articles that are related to the one being viewed. These relations are established by the Meta keywords. <br>All the keywords of the current Article are searched against all the keywords of all other published Articles. For example, you may have an Article on &quot;Breeding Parrots&quot; and another on &quot;Hand Raising Black Cockatoos&quot ...
+MOD_RELATED_XML_DESCRIPTION="This module displays other Articles that are related to the one being viewed. These relations are established by the Keywords. <br>All the keywords of the current Article are searched against all the keywords of all other published Articles. For example, you may have an Article on &quot;Breeding Parrots&quot; and another on &quot;Hand Raising Black Cockatoos&quot;. If ...
+
```
```suggestion
```

brianteeman replied:
I am doing this to make you happy but almost all language files have an extra line

brianteeman committed:
Update language/en-GB/mod_related_items.sys.ini

Co-Authored-By: Quy <quy@fluxbb.org>

brianteeman committed:
Update Content.php

brianteeman committed:
Update Content.php

Quy commented on administrator/language/en-GB/com_banners.ini:
```diff
@@ -135,7 +135,7 @@ COM_BANNERS_HEADING_COUNT_DESC="Count descending"
 COM_BANNERS_HEADING_IMPRESSIONS_ASC="Impressions ascending"
 COM_BANNERS_HEADING_IMPRESSIONS_DESC="Impressions descending"
-COM_BANNERS_HEADING_METAKEYWORDS="Meta Keywords"
+COM_BANNERS_HEADING_METAKEYWORDS="Keywords"
```
Remove as it is not being used.

brianteeman replied:
was it being used before this PR?

Quy replied:
It doesn't appear to be. I did a search on the constant and found nothing.

brianteeman committed:
Update com_banners.ini

jwaisner commented:
I have tested this item :white_check_mark: successfully on 2e575aa377a30577ada00e8547fd6548d6e49c22<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

Quy commented:
I have tested this item :white_check_mark: successfully on 2e575aa377a30577ada00e8547fd6548d6e49c22<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25258">issues.joomla.org/tracker/joomla-cms/25258</a>.</sub>

brianteeman commented:
yay!!!!

richard67 commented:
@brianteeman Today at 12 UTC:
![j4-install-persian-language-notice](https://user-images.githubusercontent.com/7413183/74084957-810d0f80-4a74-11ea-9b84-b2c6d82aed9b.png)

Content language isn't created with the language installation.

richard67 commented:
New issue or PR?

richard67 commented:
Will work on it.

richard67 commented:
Is SQL problem ... will fix it, be patient for new PR.

richard67 commented:
@brianteeman Shall the column `keywords` be removed from database table `#__languages`, or shall we keep it and use empty string default value? And how about other places where kewords were removed? I.e. what is the plan?

brianteeman commented:
the intention was not to remove the fields in this pr as you will see that the field is still used for mod_related and com_banners

richard67 commented:
@brianteeman Thanks for feedback. Yes, noticed that, too, and am in the middle of making PR for allowing empty values in db. am just writing the testing instructions.

richard67 commented:
@brianteeman Could you check and test or review or both PR #27854 ?
